<a href="https://colab.research.google.com/github/VadymBoyko/DS-HW01/blob/main/HW10_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.applications import VGG16
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, GlobalAveragePooling2D
from tensorflow.keras.optimizers import Adam

In [2]:
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

In [3]:
x_train = np.expand_dims(x_train, axis=-1)
x_test = np.expand_dims(x_test, axis=-1)

In [4]:
x_train = np.repeat(x_train, 3, axis=-1)
x_test = np.repeat(x_test, 3, axis=-1)

In [5]:
x_train = tf.image.resize(x_train, (64, 64))
x_test = tf.image.resize(x_test, (64, 64))

In [6]:
x_train = x_train / 255.0
x_test = x_test / 255.0

In [7]:
y_train = tf.keras.utils.to_categorical(y_train, num_classes=10)
y_test = tf.keras.utils.to_categorical(y_test, num_classes=10)

In [8]:
x_train.shape

TensorShape([60000, 64, 64, 3])

In [8]:
base_model = VGG16(weights='imagenet', include_top=False, input_shape=(64, 64, 3))

In [9]:
model = Sequential([
    base_model,
    GlobalAveragePooling2D(),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

In [10]:
for layer in base_model.layers:
    layer.trainable = False

In [12]:
model.compile(optimizer=Adam(learning_rate=0.001), loss='categorical_crossentropy', metrics=['accuracy'])

In [13]:
model.fit(x_train, y_train, batch_size=64, epochs=10, validation_split=0.1)

Epoch 1/10
844/844 [==============================] - 37s 30ms/step - loss: 0.5442 - accuracy: 0.8156 - val_loss: 0.4042 - val_accuracy: 0.8592
Epoch 2/10
844/844 [==============================] - 25s 29ms/step - loss: 0.3798 - accuracy: 0.8636 - val_loss: 0.3743 - val_accuracy: 0.8678
Epoch 3/10
844/844 [==============================] - 25s 29ms/step - loss: 0.3496 - accuracy: 0.8733 - val_loss: 0.3618 - val_accuracy: 0.8700
Epoch 4/10
844/844 [==============================] - 25s 30ms/step - loss: 0.3309 - accuracy: 0.8791 - val_loss: 0.3479 - val_accuracy: 0.8772
Epoch 5/10
844/844 [==============================] - 25s 30ms/step - loss: 0.3158 - accuracy: 0.8842 - val_loss: 0.3321 - val_accuracy: 0.8775
Epoch 6/10
844/844 [==============================] - 25s 30ms/step - loss: 0.3016 - accuracy: 0.8886 - val_loss: 0.3329 - val_accuracy: 0.8808
Epoch 7/10
844/844 [==============================] - 25s 30ms/step - loss: 0.2917 - accuracy: 0.8918 - val_loss: 0.3240 - val_accuracy:

In [14]:
test_loss, test_acc = model.evaluate(x_test, y_test)
print("Test accuracy:", test_acc)

313/313 [==============================] - 6s 18ms/step - loss: 0.3393 - accuracy: 0.8821
Test accuracy: 0.882099986076355


In [11]:
for layer in base_model.layers[-4:]:
    layer.trainable = True

In [12]:
model.compile(optimizer=Adam(learning_rate=0.001), loss='categorical_crossentropy', metrics=['accuracy'])

In [13]:
model.fit(x_train, y_train, batch_size=64, epochs=10, validation_split=0.1)

Epoch 1/10
844/844 [==============================] - 47s 38ms/step - loss: 0.3893 - accuracy: 0.8604 - val_loss: 0.3131 - val_accuracy: 0.8870
Epoch 2/10
844/844 [==============================] - 31s 37ms/step - loss: 0.2407 - accuracy: 0.9122 - val_loss: 0.2836 - val_accuracy: 0.9092
Epoch 3/10
844/844 [==============================] - 31s 37ms/step - loss: 0.2077 - accuracy: 0.9251 - val_loss: 0.2256 - val_accuracy: 0.9183
Epoch 4/10
844/844 [==============================] - 31s 37ms/step - loss: 0.1837 - accuracy: 0.9327 - val_loss: 0.2150 - val_accuracy: 0.9233
Epoch 5/10
844/844 [==============================] - 32s 37ms/step - loss: 0.1619 - accuracy: 0.9413 - val_loss: 0.2230 - val_accuracy: 0.9227
Epoch 6/10
844/844 [==============================] - 31s 37ms/step - loss: 0.1482 - accuracy: 0.9471 - val_loss: 0.2186 - val_accuracy: 0.9245
Epoch 7/10
844/844 [==============================] - 32s 38ms/step - loss: 0.1367 - accuracy: 0.9511 - val_loss: 0.2122 - val_accuracy:

In [14]:
test_loss, test_acc = model.evaluate(x_test, y_test)
print("Test accuracy:", test_acc)

313/313 [==============================] - 7s 18ms/step - loss: 0.2731 - accuracy: 0.9255
Test accuracy: 0.9254999756813049


**ПІДСУМКИ:**

трохи грався ресайзом, коли робив ресайз до 32х32, то точність була 0.86, при ресайзі до 64х64 піднялась до 0.88


Розблокувавши останній слой для донавчання точність підвищилась до 0.925


